src/wagtail_ai/context.py

```diff
@@ -0,0 +1,87 @@
+import json
+from string import Formatter
+from typing import Any, Type, cast
+
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.files import File
+from django.utils.translation import gettext as _
+from wagtail.images.models import AbstractImage
+
+
+class PromptContext(dict[str, Any]):
+    """
+    A dict-like object to interpolate a prompt string's placeholders with values.
+    Provides a ``clean()`` method to validate that all placeholders can be handled.
+
+    Validators can be registered with ``register_validator()`` to validate and
+    transform a key's value or provide a fallback.
+
+    A missing key without a registered validator is set as ``{key}`` so
+    ``str.format_map()`` leaves it as-is.
+    """
+
+    validators: dict[str, Any] = {}
+
+    @classmethod
+    def register_validator(cls, name: str, validator):
+        """
+        Register a validator function for a given placeholder key.
+
+        Validator functions should accept a single argument, the context
+        instance, and return the value to use for the key. They can raise
+        ``ValidationError`` if the key is required but invalid or missing from
+        the context.
+        """
+        cls.validators[name] = validator
+
+    def clean(self, prompt: str):
+        """
+        Validate that all placeholder keys in the prompt can be handled.
+        """
+        formatter = Formatter()
+        placeholders = [name for _, name, *_ in formatter.parse(prompt) if name]
+        for key in placeholders:
+            if key in self.validators:
+                # Provide the whole context to the validator to allow it to
+                # access other keys if needed
+                self[key] = self.validators[key](self)
```

Comment:
Would it be useful for the validators to also take the prompt as a parameter? If so, it's probably more useful to pass the `Prompt` object rather than just the prompt string to this `clean()` method and the validator.